components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/prompt/GeckoPromptDelegate.kt

```diff
@@ -238,32 +239,62 @@ internal class GeckoPromptDelegate(private val geckoEngineSession: GeckoEngineSe
             }
         }
 
-        val promptRequest = when (geckoPrompt.type) {
-            GECKO_PROMPT_CHOICE_TYPE.SINGLE -> SingleChoice(
-                choices,
-                onConfirmSingleChoice,
-                onDismiss
-            )
-            GECKO_PROMPT_CHOICE_TYPE.MENU -> MenuChoice(
-                choices,
-                onConfirmSingleChoice,
-                onDismiss
-            )
-            GECKO_PROMPT_CHOICE_TYPE.MULTIPLE -> MultipleChoice(
-                choices,
-                onConfirmMultipleSelection,
-                onDismiss
-            )
-            else -> throw InvalidParameterException("${geckoPrompt.type} is not a valid Gecko @Choice.ChoiceType")
-        }
+        val promptRequest = geckoPrompt.toPromptRequest(
+            choices,
+            onConfirmSingleChoice,
+            onConfirmMultipleSelection,
+            onDismiss
+        )
 
         geckoEngineSession.notifyObservers {
-            onPromptRequest(promptRequest)
+            onPromptRequest(
+                promptRequest.also {
+                    geckoPrompt.delegate = PromptInstanceUpdateDelegate(
+                        geckoEngineSession,
+                        it.uid
+                    ) { prompt ->
+                        when (prompt) {
+                            is PromptDelegate.ChoicePrompt ->
+                                prompt.toPromptRequest(
+                                    convertToChoices(prompt.choices),
+                                    onConfirmSingleChoice,
+                                    onConfirmMultipleSelection,
+                                    onDismiss
+                                )
+                            else -> null
+                        }
```

Comment:
This part is bit hard to grasp, could we refactor it in way that is bit simpler?

Reply:
As this behavior is common for all instances of `PromptInstanceUpdateDelegate` we could move the update part inside of `PromptInstanceUpdateDelegate`, and maybe we may keep keep a reference of the previous prompt request and just copy the values to create a new version.

Reply:
It looks like the only thing that it's changing, it's the choices options everything else is still being used from the previous prompt request and gecko prompt.

Reply:
For example :

```kotlin
internal class PromptInstanceUpdateDelegate(
    private val geckoSession: GeckoEngineSession,
    private val previousPrompt: PromptRequest,
) : GeckoSession.PromptDelegate.PromptInstanceDelegate {

    override fun onPromptUpdate(prompt: GeckoSession.PromptDelegate.BasePrompt) {
        when (prompt) {
            is ChoicePrompt -> {
                val updatedPromptRequest = updateChoices(prompt) ?: return
                geckoSession.notifyObservers {
                    this.onPromptUpdate(previousPrompt.uid, updatedPromptRequest)
                }
            }
        }
    }

    private fun updateChoices(prompt: ChoicePrompt): PromptRequest? {
        return when (previousPrompt) {
            is PromptRequest.MenuChoice -> {
                previousPrompt.copy(choices = convertToChoices(prompt.choices))
            }
            is PromptRequest.MultipleChoice -> {
                previousPrompt.copy(choices = convertToChoices(prompt.choices))
            }
            is PromptRequest.SingleChoice -> {
                previousPrompt.copy(choices = convertToChoices(prompt.choices))
            }
            else -> null
        }
    }
}

/**
* This could be shared between GeckoPromptDelegate 
 */
fun convertToChoices(
    geckoChoices: Array<out GeckoChoice>
): Array<Choice> {

    return geckoChoices.map { geckoChoice ->
        val choice = geckoChoice.toChoice()
        choice
    }.toTypedArray()
}

```

And in `GeckoPromptDelegate` we will only have: 


```kotlin
        geckoPrompt.delegate = PromptInstanceUpdateDelegate(
            geckoEngineSession,
            promptRequest
        )

        geckoEngineSession.notifyObservers {
            onPromptRequest(promptRequest)
        }
```

Reply:
Maybe `convertToChoices` could be in ` GeckoChoice.kt` to be something shared